src/core/connection_pool.c

```diff
@@ -0,0 +1,521 @@
+
+const char*
+QuicConnPoolAllocServerNameCopy(
+    _In_ const char* ServerName,
+    _In_ size_t ServerNameLength
+    )
+{
+    char* ServerNameCopy = CXPLAT_ALLOC_NONPAGED(ServerNameLength + 1, QUIC_POOL_SERVERNAME);
+        if (ServerNameCopy == NULL) {
+            QuicTraceEvent(
+                AllocFailure,
+                "Allocation of '%s' failed. (%llu bytes)",
+                "Server name",
+                ServerNameLength + 1);
+        } else {
+            CxPlatCopyMemory(ServerNameCopy, ServerName, ServerNameLength);
+            ServerNameCopy[ServerNameLength] = 0;
+        }
+
+    return ServerNameCopy;
+}
+
+_IRQL_requires_max_(PASSIVE_LEVEL)
+QUIC_STATUS
+QuicConnPoolGetStartingLocalAddress(
+    _In_ QUIC_ADDR* RemoteAddress,
+    _Out_ QUIC_ADDR* LocalAddress
+    )
+{
+    CXPLAT_SOCKET* Socket = NULL;
+    CXPLAT_UDP_CONFIG UdpConfig = { .RemoteAddress = RemoteAddress, };
+    QUIC_STATUS Status =
+        CxPlatSocketCreateUdp(MsQuicLib.Datapath, &UdpConfig, &Socket);
+    if (QUIC_SUCCEEDED(Status)) {
+        CXPLAT_DBG_ASSERT(Socket != NULL);
+        CxPlatSocketGetLocalAddress(Socket, LocalAddress);
+        CxPlatSocketDelete(Socket);
+    }
+
+    return Status;
+}
+
+_IRQL_requires_max_(PASSIVE_LEVEL)
+QUIC_STATUS
+QuicConnPoolGetInterfaceIndexForLocalAddress(
+    _In_ QUIC_ADDR* LocalAddress,
+    _Out_ uint32_t* InterfaceIndex
+    )
+{
+    CXPLAT_ADAPTER_ADDRESS* Addresses = NULL;
+    uint32_t AddressesCount;
+
+    *InterfaceIndex = 0;
+
+    QUIC_STATUS Status =
+        CxPlatDataPathGetLocalAddresses(
+            MsQuicLib.Datapath,
+            &Addresses,
+            &AddressesCount);
+    if (QUIC_SUCCEEDED(Status)) {
+
+        for (uint32_t i = 0; i < AddressesCount; i++) {
+            if (QuicAddrCompareIp(LocalAddress, &Addresses[i].Address)) {
+                *InterfaceIndex = Addresses[i].InterfaceIndex;
+                break;
+            }
+        }
+
+        if (*InterfaceIndex == 0) {
+            Status = QUIC_STATUS_NOT_FOUND;
+            QuicTraceLogError(
+                ConnPoolLocalAddressNotFound,
+                "[conp] Failed to find local address, 0x%x",
+                Status);
+        }
+
+        CXPLAT_FREE(Addresses, QUIC_POOL_DATAPATH_ADDRESSES);
+    }
+
+    return Status;
+}
+
+_IRQL_requires_max_(PASSIVE_LEVEL)
+QUIC_STATUS
+QUIC_API
+MsQuicConnectionPoolCreate(
+    _In_ QUIC_CONNECTION_POOL_CONFIG* Config,
+    _Out_writes_(Config->NumberOfConnections)
+        HQUIC* ConnectionPool
+    )
+{
+    QUIC_CONNECTION** Connections = (QUIC_CONNECTION**)ConnectionPool;
+    CXPLAT_RSS_CONFIG* RssConfig = NULL;
+    uint32_t* RssProcessors = NULL;
+    uint32_t* ConnectionCounts = NULL;
+    const char* ServerNameCopy = NULL;
+    CXPLAT_TOEPLITZ_HASH ToeplitzHash;
+    QUIC_STATUS Status = QUIC_STATUS_SUCCESS;
+    uint32_t CreatedConnections = 0;
+
+    if (Config == NULL || ConnectionPool == NULL || Config->Registration == NULL ||
```

Comment:
nit: That if test is very hard to read.
It could easily be split in several tests and potentially in a "validateConfig" function.

Reply:
I'll break it up to make it easier to read, but this is an area where I disagree with making a helper function.